fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/IndexKeyValueToPartialRecord.java

```diff
@@ -368,6 +368,12 @@ public void addRequiredMessageFields() {
                 if (fieldDescriptor.isRequired() && fieldDescriptor.getType() == Descriptors.FieldDescriptor.Type.MESSAGE) {
                     nestedBuilders.putIfAbsent(fieldDescriptor.getName(), new Builder(fieldDescriptor.getMessageType(), true));
                 }
+                if (fieldDescriptor.getType() == Descriptors.FieldDescriptor.Type.MESSAGE) {
```

Comment:
I would either do as @alecgrieser suggests below and have a recursion pass over all child builders at the end rather than field-by-field, _or_ combine this `if` and the previous one into a single "nested message" one that (1) add builder if message field itself is required and (2) recurse if there is a child builder.